File: python/test_parsers.py
#!/usr/bin/env python3
"""Correctness suite — every case is run against every strategy.

Mirrors cpp/tests/test_parsers.cpp: variable environment a=2, b=3, c=4, d=5,
plus a few numeric-literal cases for features a-d cannot express. Exit 0 on
success, 1 on any failure (so it works as a CI/ctest-style check).
"""
import math
import sys

sys.path.insert(0, __file__.rsplit("/", 1)[0])
from mathparser import all_evaluators  # noqa: E402

VARS = [0.0] * 26
VARS[0], VARS[1], VARS[2], VARS[3] = 2.0, 3.0, 4.0, 5.0  # a,b,c,d

CASES = [
    ("a + b * c", 14),
    ("(a + b) * c", 20),
    ("a * (b + c) - d", 9),
    ("(a + b) * b / d", 3),
    ("-(b + c) * a", -14),
    ("d - a - b", 0),
    ("d * c / a", 10),
    ("a ^ a ^ a", 16),
    ("-a ^ a", -4),
    ("a ^ -b", 0.125),
    ("3.5 * a + 1.5", 8.5),
    ("2e2 + a", 202),
    ("--a", 2),
    ("-a * b", -6),
    ("+d - +a", 3),
    # IEEE special values — must match C++ std::pow / IEEE division exactly
    ("0 ^ -1", math.inf),
    ("1 / (0 * -1)", -math.inf),
    ("(0 - 1e155) ^ 3", -math.inf),
    ("(0 / 0) / 0", math.nan),
    ("(0 - a) ^ 0.5", math.nan),
    # literal overflow/underflow is a value (inf / 0), not a syntax error
    ("1e400", math.inf),
    ("1e-400 + a", 2),
]

ERRORS = [
    "a +", "(a + b", "a b", "* a", "a + * b", "",
    # stray ')' and adjacent operand groups (regressions: the shunting-yard
    # family accepted these), and a digitless number
    "a)", "(a)(b)", "a(3)", ".",
]


def nearly(a, b):
    if math.isnan(b):
        return math.isnan(a)
    if a == b:  # covers ±inf
        return True
    return abs(a - b) <= 1e-9 * max(1.0, abs(a), abs(b))


def main():
    checks = failures = 0
    evs = all_evaluators()
    for ev in evs:
        for expr, expected in CASES:
            checks += 1
            try:
                got = ev.eval(expr, VARS)
                if not nearly(got, expected):
                    print(f"FAIL [{ev.name:<26}] {expr!r} = {got:g}, expected {expected:g}")
                    failures += 1
            except Exception as e:  # noqa: BLE001
                print(f"FAIL [{ev.name:<26}] {expr!r} threw: {e}")
                failures += 1
        for expr in ERRORS:
            checks += 1
            try:
                ev.eval(expr)
                print(f"FAIL [{ev.name:<26}] {expr!r} should have raised")
                failures += 1
            except Exception:  # noqa: BLE001
                pass  # expected

    print(f"\n{checks} checks across {len(evs)} evaluators, {failures} failure(s)")
    return 1 if failures else 0


if __name__ == "__main__":
    sys.setrecursionlimit(200_000)
    sys.exit(main())

File: cpp/tests/test_parsers.cpp
// Dependency-free correctness suite. Every case is run against every evaluation
// strategy so they are held to an identical specification.
#include "parser/evaluator.hpp"
#include "parser/reeval.hpp"

#include "test_util.hpp"

#include <array>
#include <cmath>
#include <limits>
#include <print>
#include <exception>
#include <string>
#include <string_view>
#include <vector>

using namespace mp;

namespace {

int g_checks = 0;
int g_failures = 0;

struct Case    { std::string_view expr; double expected; };
struct ErrCase { std::string_view expr; };

using mp::test::nearly;

void checkValue(IEvaluator& ev, const Case& c, const double* vars = nullptr) {
    ++g_checks;
    try {
        const double got = ev.eval(c.expr, vars);
        if (!nearly(got, c.expected)) {
            std::println("FAIL [{:<26}] \"{}\" = {:g}, expected {:g}",
                        ev.name(), c.expr, got, c.expected);
            ++g_failures;
        }
    } catch (const std::exception& e) {
        std::println("FAIL [{:<26}] \"{}\" threw: {}",
                    ev.name(), c.expr, e.what());
        ++g_failures;
    }
}

void checkError(IEvaluator& ev, const ErrCase& c) {
    ++g_checks;
    try {
        (void)ev.eval(c.expr);
        std::println("FAIL [{:<26}] \"{}\" should have thrown",
                    ev.name(), c.expr);
        ++g_failures;
    } catch (const std::exception&) {
        // expected
    }
}

}  // namespace

int main() {
    // Variable environment: a=2, b=3, c=4, d=5.
    const auto env = mp::test::testEnv();
    const double* vars = env.data();

    // All cases exercise the variable code path; a few numeric literals remain
    // for features that cannot be expressed through a-d (scientific notation,
    // unary stacking on a literal, etc.).
    const std::vector<Case> cases = {
        // precedence: a=2, b=3, c=4  →  a+b*c = 2+12 = 14
        {"a + b * c",            14},
        // parentheses override precedence: (a+b)*c = 5*4 = 20
        {"(a + b) * c",          20},
        // mixed: a*(b+c)-d = 2*7-5 = 9
        {"a * (b + c) - d",       9},
        // nested parens: (a+b)*b/d = 5*3/5 = 3
        {"(a + b) * b / d",       3},
        // unary negation: -(b+c)*a = -7*2 = -14
        {"-(b + c) * a",        -14},
        // left-associativity of -: d-a-b = 5-2-3 = 0
        {"d - a - b",             0},
        // left-associativity of /: d*c/a = 20/2 = 10
        {"d * c / a",            10},
        // right-associative ^: a^a^a = 2^(2^2) = 2^4 = 16
        {"a ^ a ^ a",            16},
        // unary/power interplay: -a^a = -(2^2) = -4
        {"-a ^ a",               -4},
        // negative exponent: a^-b = 2^-3 = 0.125
        {"a ^ -b",            0.125},
        // numeric literal + variable: 3.5*a+1.5 = 7+1.5 = 8.5
        {"3.5 * a + 1.5",       8.5},
        // scientific notation literal: 2e2+a = 200+2 = 202
        {"2e2 + a",             202},
        // unary stacking on a literal
        {"--a",                   2},
        // unary on variable: -a*b = -6
        {"-a * b",               -6},
        // unary plus: +d-+a = 5-2 = 3
        {"+d - +a",               3},
        // IEEE special values — pow/div corners shared across the languages
        {"0 ^ -1",     std::numeric_limits<double>::infinity()},
        {"1 / (0 * -1)", -std::numeric_limits<double>::infinity()},
        {"(0 - 1e155) ^ 3", -std::numeric_limits<double>::infinity()},
        {"(0 / 0) / 0", std::numeric_limits<double>::quiet_NaN()},
        {"(0 - a) ^ 0.5", std::numeric_limits<double>::quiet_NaN()},
        // literal overflow/underflow is a value (inf / 0), not a syntax error
        {"1e400", std::numeric_limits<double>::infinity()},
        {"1e-400 + a", 2},
    };

    const std::vector<ErrCase> errs = {
        {"a +"}, {"(a + b"}, {"a b"}, {"* a"}, {"a + * b"}, {""},
        // stray ')' and adjacent operand groups (regression: bytecode-vm and
        // the re-eval compilers accepted these), and a digitless number
        {"a)"}, {"(a)(b)"}, {"a(3)"}, {"."},
    };

    auto evs = all_evaluators();
    for (auto& ev : evs) {
        for (const auto& c : cases) checkValue(*ev, c, vars);
        for (const auto& e : errs)  checkError(*ev, e);
    }

    // Parallel variants: same spec, plus malformed inputs long enough to fork
    // (regression: pool*/dfork* used to std::terminate / hang when a subtree
    // parsed on a worker thread threw).
    std::string longOk, longTrail, longParen;
    for (int i = 0; i < 400; ++i) {
        longOk += i ? " + a" : "a";
    }
    longTrail = longOk + " +";
    longParen = longOk; longParen[longParen.size() / 2] = ')';
    const Case longCase{longOk, 800};
    const std::vector<ErrCase> longErrs = {{longTrail}, {longParen}};
    auto pevs = parallel_evaluators();
    for (auto& ev : pevs) {
        for (const auto& c : cases) checkValue(*ev, c, vars);
        for (const auto& e : errs)  checkError(*ev, e);
        checkValue(*ev, longCase, vars);
        for (const auto& e : longErrs) checkError(*ev, e);
    }

    // Re-eval compilers — reuse the same env defined above.
    const std::vector<Case> varCases = {
        {"a + b * c", 14},
        {"(a + b) * c", 20},
        {"a ^ 2 + b", 7},
        {"-a + b", 1},
        {"d / b - a", 5.0 / 3 - 2},
        {"2 + 3 * 4", 14},  // constant still works through compilers
    };
    auto compilers = all_compilers();
    for (auto& comp : compilers) {
        for (const auto& e : errs) {
            ++g_checks;
            try {
                (void)comp->compile(e.expr)->eval(vars);
                std::println("FAIL [{:<26}] \"{}\" should have thrown",
                            comp->name(), e.expr);
                ++g_failures;
            } catch (const std::exception&) {
                // expected
            }
        }
        for (const auto& vc : varCases) {
            ++g_checks;
            try {
                const double got = comp->compile(vc.expr)->eval(vars);
                if (!nearly(got, vc.expected)) {
                    std::println("FAIL [{:<26}] \"{}\" = {:g}, expected {:g}",
                                comp->name(), vc.expr, got, vc.expected);
                    ++g_failures;
                }
            } catch (const std::exception& e) {
                std::println("FAIL [{:<26}] \"{}\" threw: {}",
                            comp->name(), vc.expr, e.what());
                ++g_failures;
            }
        }
    }

    std::println("\n{} checks across {} evaluators + {} parallel variants + {} compilers, {} failure(s)",
                g_checks, evs.size(), pevs.size(), compilers.size(), g_failures);
    return g_failures == 0 ? 0 : 1;
}
